internal/era2/builder.go

```diff
@@ -0,0 +1,339 @@
+// TotalDifficulty is little-endian encoded.
+
+// AccumulatorRoot is only defined for epochs with pre-merge data.
+// HeaderRecord is defined in the Portal Network specification[^5].
+
+// BlockIndex stores relative offsets to each compressed block entry. The format is:
+
+//    block-index := starting-number | index | index | index ... | count
+// All values in the block index are little-endian uint64.
+
+// starting-number is the first block number in the archive. Every index is a defined relative to index's location in the file. The total number of block entries in the file is recorded in count.
+
+// Due to the accumulator size limit of 8192, the maximum number of blocks in an Era batch is also 8192. This is also the value of SLOTS_PER_HISTORICAL_ROOT[^6] on the Beacon chain, so it is nice to align on the value.
+
+import (
+	"bytes"
+	"encoding/binary"
+	"errors"
+	"fmt"
+	"io"
+	"math/big"
+
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/ethereum/go-ethereum/core/types"
+	"github.com/ethereum/go-ethereum/internal/era/e2store"
+	"github.com/ethereum/go-ethereum/rlp"
+	"github.com/golang/snappy"
+)
+
+// Type constants for the e2store entries in the Era2 format.
+const (
+	TypeVersion            uint16 = 0x3265
+	TypeCompressedHeader   uint16 = 0x08
+	TypeCompressedBody     uint16 = 0x09
+	TypeCompressedReceipts uint16 = 0x0a
+	TypeTotalDifficulty    uint16 = 0x0b
+	TypeProof              uint16 = 0x0c
+	TypeAccumulatorRoot    uint16 = 0x0d
+	TypeBlockIndex         uint16 = 0x3267
+	MaxEraESize            int    = 8192
+	headerSize             uint64 = 8
+)
+
+// Temporary buffer for writing blocks until the Finalize method is called.
+type buffer struct {
+	headers  [][]byte
+	bodies   [][]byte
+	receipts [][]byte
+	proofs   [][]byte
+	tds      []*big.Int
+}
+
+// The offsets holds the offsets of the different block components in the e2store file. Eventually these offsets will be used to write the index table at the end of the file.
+type offsets struct {
+	headers      []uint64
+	bodys        []uint64
+	receipts     []uint64
+	proofoffsets []uint64
+	tdoff        []uint64
+}
+
+// Builder is used to build an Era2 e2store file. It collects block entries and writes them to the underlying e2store.Writer.
+type Builder struct {
+	w   *e2store.Writer
+	buf *bytes.Buffer
+
+	buff buffer
+	off  offsets
+
+	hashes        []common.Hash
+	startNum      *uint64
+	written       uint64
+	expectsProofs bool
+}
+
+// NewBuilder returns a new Builder instance.
+func NewBuilder(w io.Writer) *Builder {
+	buf := bytes.NewBuffer(nil)
+	return &Builder{
+		w:   e2store.NewWriter(w),
+		buf: buf,
+	}
+}
+
+// Add writes a block entry, its reciepts, and optionally its proofs as well into the e2store file.
+func (b *Builder) Add(header types.Header, body types.Body, receipts types.Receipts, td *big.Int, proof Proof) error {
+	if len(b.buff.headers) == 0 { // first block determines wether proofs are expected
+		b.expectsProofs = proof != nil
```

Comment:
i don't think you need to track this explicitly, just check if `b.buff.proofs != nil` and if `proof == nil` or vice versa. only special case is first block, which you allow for the `b.buff.proofs` to be `nil` even if proof is non-`nil`.